### PR TITLE
Added support for Docker, GitHub Actions and graceful shutdown

### PR DESCRIPTION
**Pull request by ad-m**

### CODE REVIEW

**Comment by ad-m** [comment]
Related #624 (missing setup login and secrets of registry), #627 (missing cleanup TravisCI)

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	npm run eslint
+	npm run prettier-check
+
+.PHONY: build
+build:
+	npm install
+
+.PHONY: test
+test:
+	npm run test
```
Comment (fishcharlie) [inline Makefile:15]: Not quite sure I understand the purpose for why we need a `Makefile` for these cases. What is the benefit of running `make test` vs `npm run test`? Doesn't really seem like an advantage to me.

File: .github/workflows/javascript.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    steps:
+    - uses: actions/checkout@v1
+    - name: Set up Node
+      uses: actions/setup-node@v1
+      with:
+        node-version: '${{ matrix.node }}'
+    - name: Build
+      run: make build
+    - name: Test
+      run: make test
```
Comment (fishcharlie) [inline .github/workflows/javascript.yaml:22]: Wouldn't we want another job to run the `lint` as well?

**Review by fishcharlie — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (kherock) [reply Makefile:15]: At the risk of sounding unconstructive, I was also a little put off by the use of Makefile. I do think most of these would function fine as npm scripts as this is a relatively simple JavaScript library.

Comment (ad-m) [reply Makefile:15]: I personally don't like have non-JavaScript things like Docker in `package.json`. `package.json` does not support variables what is useful for build image.

However, I am open to cooperation and taking into account the suggested changes. I will drop `Makefile` and create a new PR.

**Review by kherock — COMMENTED** [review]
I'm not totally against the Makefile for things like building a Docker image, though I'm pretty comfortable typing in `docker build . -t jamhall/s3rver` myself anytime I want to build a fresh local copy.

I've opened #681, and I'd appreciate your feedback there @ad-m. If you any serious grievances with it let me know, otherwise I think I'd like to repurpose this PR as simply "Official Docker support".

**Pull request closed without merging**